### PR TITLE
CB-8054 Support taking pictures from file for WP8

### PR DESCRIPTION
**Pull request by muratsu**
Add missing take picture from file capability for WP8

### CODE REVIEW

**Comment by vladimir-kotikov** [comment]
LGTM

File: src/windows/CameraProxy.js [hunk 1]
```diff
@@ -134,16 +134,84 @@ function resizeImageBase64(successCallback, errorCallback, file, targetWidth, ta
...
+        }
+    };
+
+    var fileOpenPicker = new Windows.Storage.Pickers.FileOpenPicker();
+    fileOpenPicker.suggestedStartLocation = Windows.Storage.Pickers.PickerLocationId.documentsLibrary;
+    if (mediaType == Camera.MediaType.PICTURE) {
+        fileOpenPicker.fileTypeFilter.replaceAll([".png", ".jpg", ".jpeg"]);
+    }
+    else if (mediaType == Camera.MediaType.VIDEO) {
+        fileOpenPicker.fileTypeFilter.replaceAll([".avi", ".flv", ".asx", ".asf", ".mov", ".mp4", ".mpg", ".rm", ".srt", ".swf", ".wmv", ".vob"]);
```
Comment (nikhilkh) [inline src/windows/CameraProxy.js]: Wonder if this list of video formats should be shared with the one from windows below? It's amazing that so many video formats are supported here.

**Comment by nikhilkh** [comment]
Is there a docs update needed - now that this is supported?

**Comment by muratsu** [comment]
The docs didn't mention this as a quirk, so no doc updates.
I've also updated the phone video codec list.

**Comment by nikhilkh** [comment]
LGTM - Please rebase for a merge

**Comment by muratsu** [comment]
This one is ready to be merged. Would be glad if someone can merge this in.

**Comment by ekambarrao** [comment]
hi all, is this problem resolved for windows phone universal. I am not able to use this for accessing photolibrary. Also camera is not working as it is with windows phone 8. Your quick reply is very helpful to me.

**Comment by muratsu** [comment]
Yes, this is resolved for windows phone universal. 
What problem are you running into with winphone8? You can always open a [jira](https://issues.apache.org/jira/browse/CB-9074?jql=project%20%3D%20CB) for us to investigate. The more information you provide, faster we can resolve the issue.

**Comment by ekambarrao** [comment]
I am having the problem of "not Supported" error. Would you suggest me to reinstall the plugin?

**Comment by muratsu** [comment]
Make sure you've version 1.1.0 (May 06, 2015). You can check installed plugin version by executing `cordova plugin` on your project root.

**Comment by ekambarrao** [comment]
Thanks for quick reply. can you pls share me the url link of camera plugin! I am using org.apache.cordova.camera plugin of visual studio and seems that I am outdated here with this.

**Comment by muratsu** [comment]
You can get the latest (edge) from [github](https://github.com/apache/cordova-plugin-camera.git) however I suggest you clone the github repo and checkout 19feee9cb049 for version 1.1.0. Releases are tested more thoroughly than edge.

**Comment by ekambarrao** [comment]
Hi Muratsu, thanks for quick reply again. I was good with camera but bad with gallery. Now good with Gallery but bad with camera. It is breaking with error "Camera Initialization error TypeError: getAvailableMediaStreamProperties: 'this' is not an Inspectable Object". Can you suggest the solution please! thx

**Comment by muratsu** [comment]
That just got fixed today. Please update :)

**Comment by ekambarrao** [comment]
yep...now it is working perfect. however I am looking for one more suggestion about, how to cancel the camera!!

**Comment by ekambarrao** [comment]
It is working perfect now thx.

**Comment by ekambarrao** [comment]
muratsu, I have footer in the page where I activate the camera, camera has "cancel" but it is going under the footer. How can I control the size of the camera screen??

**Comment by ekambarrao** [comment]
Hi, I am back....I used this earlier and was working perfectly alright. The same code is not working now. Camera is abending when saving to library on phone..Gallery lists the library and abends....kicks out of application. i reimplemented the plugin, no use. I have downloaded the 1.2.1, but no use. Any idea what could be the reason!!? Environment is Windows 8.1 phone universal.

**Comment by muratsu** [comment]
Hmm, can you share your camera settings with me? I've tried the following with ee5cfe89 (master):
```
        destinationType: Camera.DestinationType.FILE_URI,
        encodingType: Camera.EncodingType.JPEG,
        sourceType : Camera.PictureSourceType.PHOTOLIBRARY
```
```
        destinationType: Camera.DestinationType.FILE_URI,
        encodingType: Camera.EncodingType.JPEG,
        sourceType : Camera.PictureSourceType.CAMERA,
        saveToPhotoAlbum: true
```

both of them working fine for me (win-phone 8.1).

**Comment by ekambarrao** [comment]
I am not sure what is the reason, my code is working fine, plugin is working fine, but after the closure  of execution of returning the image to page of App, Android 19 with Cordova 3.6.4, after the execution, cordova is executing further that is killing the app. Where the same may be true in other platforms as well. What is your suggestion in the platform versions, what should I be using?

**Comment by ekambarrao** [comment]
I just reinstalled Android and Cordova, what the difference in behavior i have seen is, earlier, it was exiting in nano seconds, now it is showing the library and waiting there for a second in a no response state and then exiting.

**Comment by ekambarrao** [comment]
Muratsu, Can you suggest me a solution for this as my app has been held back only due to this!!!.

**Comment by ekambarrao** [comment]
Muratsu, please find the following videos where you can see the problems with this plugin.

works in windows emulator.

https://www.youtube.com/watch?v=e17Dhujnm1A&feature=youtu.be

android device failure, not able to record the emulator, but it is also failing there. 

https://www.youtube.com/watch?v=Y7cQHc3i5Ew&feature=youtu.be

Windows device failure

https://www.youtube.com/watch?v=LLofT2ghxIs&feature=youtu.be

What figure out is, it is not retaining the app state, earlier it was not that way, do i need to explicitly maintain the app state?

If you can't access let me know.

**Comment by ekambarrao** [comment]
As I was doubtful, the issue was related to App State Maintenance. Some how it has been poorly maintained in the app. This experience and solution might be useful to others.

**Comment by purplecabbage** [comment]
You can try this: 
https://msdn.microsoft.com/en-us/library/windows/apps/jj735579(v=vs.105).aspx
 

To enable Fast Resume for your app, add the ActivationPolicy attribute to the DefaultTask element in WMAppManifest.xml and set the value to “Resume”. For this task, you need to edit the app manifest directly instead of using the manifest editor. To do this, right-click WMAppManifest.xml, click Open with, and then choose XML (Text) Editor.

**Comment by ekambarrao** [comment]
@purplecabbage , mine is cordova javascript application. I am really in a very bad shape due to lack of support from all OEM of these techs. Very bad!!!! Its been 500 hours of trials after trails to figure out how to make this work consistent way without fail but could not resolve this. Stackoverflow did not help, slack did not help.....what is the wrong .....where....god only knows....

**Comment by ekambarrao** [comment]
I used a cordova navigation sample. 

worked perfectly alright up to some extent. 

When I click on the camera, it perfectly returns to library, 

click save, 

Important Point: "it returns to page then tries to render the image"

I started making changes line by line and was testing it......to see if this makes difference or that makes difference...it broke at one point...

i rolled back changes...but it was not rolling back....

Does this have any thing to do cordovaApp_temporarykey.pfx?

do we need to have this file in the solution to have this work?

when i create another application with the same lines of code it was not working......what is happening guys....cordova is really very bad in managing its state or I am using winjs for navigation, it is bad in managing the state....it is not hitting the resume event in index.js at all.... to capture the reason for failure.....

**Comment by purplecabbage** [comment]
Please post your non-working example, and I will look at it.
Speculation is dangerous.

This has nothing to do with cordovaApp_temporarykey.pfx
Define 'rolled back changes'... via version control? did you run 'cordova prepare'?

**Comment by ekambarrao** [comment]
@purplecabbage this is like light in dark....your offer itself is like a great savior to me....:-)

I have uploaded the solutions @ the following onedrive

https://onedrive.live.com/?id=463D607901FE1563%21114&cid=463D607901FE1563&group=0

blankcordovaapp1 is failing where I experimented.

javascript(2) has been working fine.

they were in VS 2013 and it is picking up cordova 3.7.1

I have moved the code to VS 2015 and I selected 5.1.1 s platform but it is showing 4.0 folder in .cordova folder.

2015 working as long as the device in debug when disconnected again the same.

**Comment by purplecabbage** [comment]
I don't see anything at the url you posted.



My team is hiring!
@purplecabbage
risingj.com

**Comment by ekambarrao** [comment]
<iframe src="https://onedrive.live.com/embed?cid=463D607901FE1563&resid=463D607901FE1563%21114&authkey=ADmtINFmJhqVSBI" width="165" height="128" frameborder="0" scrolling="no"></iframe>

**Comment by purplecabbage** [comment]
Okay, it appears you are doing some magic with the Visual Studio TACO ( tools for apache cordova )
You will need to bring this up with the visual studio team, as ultimately what you are running is NOT strictly a cordova project.  
I was able to run your javascript(2) version as posted, but the BlankCordovaApp1 app was missing multiple files and could not be built.
Ultimately this is not the place to have this discussion, as it has nothing to do with this pull request.
If you need to discuss this further, please open a jira issue, or contact me on the cordova slack channel.
to join :
http://slack.cordova.io/ 
to view :
https://cordova.slack.com/

I am @purplecabbage everywhere.

**Comment by nikhilkh** [comment]
I agree with @purplecabbage that this is not the correct forum for this. Please use stack overflow and mark it with visual studio cordova. People from the product team look at that forum and will respond to you.

**Comment by ekambarrao** [comment]
Thanks guys...for taking time on this....

**Comment by ekambarrao** [comment]
i have a question...do i need to capture the state of the app while capturing the photo or gone to library? or the plugin will take care of that?

**Pull request merged**